plugins/out_doris/doris.c

```diff
@@ -0,0 +1,545 @@
+                out_ret = FLB_RETRY;
+                goto free_buf;
+            }
+
+            root = result.data;
+            if (root.type != MSGPACK_OBJECT_MAP) {
+                out_ret = FLB_RETRY;
+            }
+
+            for (i = 0; i < root.via.map.size; i++) {
+                msg_key = root.via.map.ptr[i].key;
+                if (msg_key.type != MSGPACK_OBJECT_STR) {
+                    out_ret = FLB_RETRY;
+                    break;
+                }
+
+                if (msg_key.via.str.size == 6 && strncasecmp(msg_key.via.str.ptr, "Status", 6) == 0) {
+                    msg_val = root.via.map.ptr[i].val;
+                    if (msg_val.type != MSGPACK_OBJECT_STR) {
+                        out_ret = FLB_RETRY;
+                        break;
+                    }
+
+                    if (msg_val.via.str.size == 7 && strncasecmp(msg_val.via.str.ptr, "Success", 7) == 0) {
+                        out_ret = FLB_OK;
+                        break;
+                    }
+                    
+                    if (msg_val.via.str.size == 15 && strncasecmp(msg_val.via.str.ptr, "Publish Timeout", 15) == 0) {
+                        out_ret = FLB_OK;
+                        break;
+                    }
+                    
+                    out_ret = FLB_RETRY;
+                    break;
+                }
+            }
+free_buf:
+            flb_free(out_buf);
+            msgpack_unpacked_destroy(&result);
+parse_done:;
+        }
+        else {
+            out_ret = FLB_RETRY;
+        }
+    }
+    else {
+        flb_plg_error(ctx->ins, "could not flush records to %s:%i (http_do=%i)",
+                      ctx->host, ctx->port, ret);
+        out_ret = FLB_RETRY;
+    }
+
+    /* cleanup */
+    
+    /*
+     * If the payload buffer is different than incoming records in body, means
+     * we generated a different payload and must be freed.
+     */
+    if (payload_buf != body) {
+        flb_free(payload_buf);
+    }
+
+    /* Destroy HTTP client context */
+    flb_http_client_destroy(c);
+
+    /* Release the TCP connection */
+    flb_upstream_conn_release(u_conn);
+
+    return out_ret;
+}
+
+static int compose_payload(struct flb_out_doris *ctx,
+                           const void *in_body, size_t in_size,
+                           void **out_body, size_t *out_size,
+                           struct flb_config *config)
+{
+    flb_sds_t encoded;
+
+    *out_body = NULL;
+    *out_size = 0;
+
+    encoded = flb_pack_msgpack_to_json_format(in_body,
+                                              in_size,
+                                              ctx->out_format,
+                                              FLB_PACK_JSON_DATE_EPOCH,
+                                              ctx->date_key,
+                                              config->json_escape_unicode);
+    if (encoded == NULL) {
+        flb_plg_error(ctx->ins, "failed to convert json");
+        return FLB_ERROR;
```

Comment:
These constants are used for coroutine results, for internal functions you can use any constants you want, the usual `-1, 0, 1` are allowed.